zookeeper-client/zookeeper-client-c/src/zookeeper.c

```diff
@@ -3961,6 +3961,19 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
 /*---------------------------------------------------------------------------*
  * ASYNC API
  *---------------------------------------------------------------------------*/
+
+/* make an attempt to send queued requests immediately without blocking */
+static int nonblocking_send(zhandle_t *zh, int rc)
+{
+    if (adaptor_send_queue(zh, 0) < 0) {
+        if (is_connected(zh)) {
+            close_zsock(zh->fd);
+            zh->state = ZOO_NOTCONNECTED_STATE;
+        }
+    }
+    return (rc < 0) ? ZMARSHALLINGERROR : ZOK;
```

Comment:
@eolivelli wrote:

> in the previous version of this code we were basically always calling `adaptor_send_queue`, now in case of previous error (rc < 0) we are not calling that function anymore.
> 
> can this have a behavioural impact ?

Good point, indeed.  I don't think we have a way to be sure, so I prefer this version.

Now… sorry to be annoying, but:

I'm not sure I like the `is_connected(zh)` condition.  You previously had `zh->fd->sock != -1` here, which seems to make more sense—as the client might *not* be connected when this function is called.  (That can happen, e.g., if the client is in the middle of a `prime_connection` or a SASL authentication sequence; applications do not always wait for a state notification before they start issuing requests.)

WDYT?

Reply:
Sounds reasonable.  Will return to using `zh->fd != -1` as the guard for calling `close_zsock()` ; new version bc8adb0b4 is pushed.